XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+manually passed to the client's constructor and propagated to the `init()`'s
+`parentId`.
+
+The constructed `handle` is passed to `init()`. The structure of any object
+passed to `init()` has no guarantee of stability, even through patch updates.
+It's meant to be used for debugging the application when additional, and
+specific, information is needed. Make sure not to hold a reference to this
+indefinitely or else the GC won't be able to collect it after node has removed
+its own reference to the handle.
+
+
+#### `before(id)`
+
+* `id` {Number}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times.
+
+
+#### `after(id)`
+
+* `id` {Number}
+
+Called immediately after the return callback is completed. If a callback throws
+but is not caught then the process will exit immediately without calling
+`after()`.
+
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In several cases this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be called manually when the TCP connection is no
+longer in need of it. Every subsequent use of a shared resource will have a new
+unique id.
+
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+`AsyncWrap` API so that all the appropriate callbacks are called. To
+accommodate this both a C++ and JS API is provided.
+
+
+### Native API
+
+```cpp
+// Helper class users can inherit from, but is not necessary. It will
+// automatically call all four callbacks.
+class AsyncHook {
+  public:
+    AsyncHook(Local<Object> handle, const char* name);
+    ~AsyncHook();
+    MaybeLocal<Value> MakeCallback(
+        const Local<Function> callback,
+        int argc,
+        Local<Value>* argv);
+    Local<Object> handle();
+    uint64_t get_uid();
+  private:
+    AsyncHook();
+    Persistent<Object> handle_;
+    const char* name_;
+    uint64_t uid_;
+}
+
+// Returns the id of the current execution context. If the return value is
+// zero then no execution has been set. This will happen if the user handles
+// I/O from native code.
+uint64_t node::GetCurrentId();
+
+// If the native API doesn't inherit from the helper class then the callbacks
+// must be triggered manually. This triggers the init() callback. The return
+// value is the uid assigned to the handle.
+// TODO(trevnorris): This always needs to be called so that the handle can be
+// placed on the Map for future query.
+uint64_t node::EmitAsyncHandleInit(Local<Object> handle);
```

Comment:
The last sentence here:
```
// must be triggered manually. This triggers the init() callback. The return
// value is the uid assigned to the handle.
```
Though the `TODO` gets in the way of seeing that. The `TODO` will be removed before this lands.